src/Npgsql/NpgsqlParameterCollection.cs

```diff
@@ -351,35 +372,42 @@ public override int IndexOf(string parameterName)
             if (_lookup is null)
                 BuildLookup();
 
-            if (_lookup!.TryGetValue(parameterName, out var indexCs))
+            if (TwoPassCompatMode && _caseSensitiveLookup!.TryGetValue(parameterName, out var indexCs))
                 return indexCs;
-
-            if (CaseInsensitiveCompatMode && _caseInsensitiveLookup!.TryGetValue(parameterName, out var indexCi))
+                
+            if (_lookup!.TryGetValue(parameterName, out var indexCi))
                 return indexCi;
 
             return -1;
         }
 
+        // Start with case-sensitive search in two pass mode.
+        if (TwoPassCompatMode)
+        {
+            for (var i = 0; i < InternalList.Count; i++)
+            {
+                var name = InternalList[i].TrimmedName;
+                if (string.Equals(parameterName, InternalList[i].TrimmedName))
```

Comment:
Can also do ReferenceEquals (like below) to make two-pass slightly more efficient

Reply:
I remember this has something to do with string.Equals going straight into reference equals while string.Equals($,$, comparison) first checks comparison validity before doing that.